### PR TITLE
Add C++ coverage to CI

### PR DESCRIPTION
**Pull request by chaeyeunpark**
Add C++ code coverage to GH Action.

### CODE REVIEW

**Comment by chaeyeunpark** [comment]
Hi all, this PR adds C++ coverage to CodeCov. Now it only uploads results from Linux and Windows, but macOS can be also added if it is necessary. Still, it seems like the coverage results only count the lines that are visible under the given environment, i.e. `#ifdef __GCC__` block is not added to the denominator of the percentage it calculates when MSVC is used.

File: .github/workflows/tests_linux.yml [hunk 1]
```diff
@@ -98,6 +117,21 @@ jobs:
           check_name: Test Report (C++) on Ubuntu
           files: Build/tests/results/report.xml
 
+      - name: Build and run unit tests for code coverage
+        run: |
+            cmake pennylane_lightning/src -BBuildCov  -DCMAKE_BUILD_TYPE=Debug -DBUILD_TESTS=ON -DENABLE_COVERAGE=ON
```
Comment (mlxd) [inline .github/workflows/tests_linux.yml:122]: Do we need to enable `-DENABLE_BLAS=ON` here also?

File: .github/workflows/tests_linux.yml [hunk 2]
```diff
@@ -30,13 +30,15 @@ jobs:
           python-version: '3.7'
 
       - uses: actions/checkout@v2
+        with:
+          fetch-depth: 2 # for codecov
```
Comment (mlxd) [inline .github/workflows/tests_linux.yml:34]: Nice -- wasn't aware this is an option.

**Review by mlxd — COMMENTED** [review]
Awesome work @chaeyeunpark 
Left a few small comments/questions.

File: pennylane_lightning/src/util/RuntimeInfo.cpp [hunk 3]
```diff
@@ -14,17 +14,15 @@
...
-#elif defined(_MSC_VER)
+#elif defined(_MSC_VER) && defined(_WIN64)
 #include <array>
 #include <intrin.h>
 #endif
-#endif
 
 namespace Pennylane::Util {
-#if defined(__x86_64__) && (defined(__GNUC__) || defined(__clang__))
+#if (defined(__GNUC__) || defined(__clang__)) && defined(__x86_64__)
```
[the remark this answers is not in the document]
Comment (chaeyeunpark) [reply pennylane_lightning/src/util/RuntimeInfo.cpp:25]: Yes but as it also support `cpuid.h` normally, I think it is fine (also tested locally).

File: .github/workflows/tests_windows.yml [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+          choco install OpenCppCoverage -y
+          echo "C:\Program Files\OpenCppCoverage" >> $env:GITHUB_PATH
+
+      - name: Build and run unit tests for code coverage
+        run: |
+            cmake ./pennylane_lightning/src -BBuild  -DBUILD_TESTS=ON -DENABLE_OPENMP=OFF -DENABLE_WARNINGS=OFF
+            cmake --build ./Build --config Debug
+            mkdir -p ./Build/tests/results
+            .\Build\tests\Debug\runner.exe --order lex --reporter junit --out .\Build\tests\results\report.xml
+            OpenCppCoverage --sources pennylane_lightning\src --export_type cobertura:coverage.xml Build\tests\Debug\runner.exe
```
[the remark this answers is not in the document]
Comment (chaeyeunpark) [reply .github/workflows/tests_windows.yml:40]: Yes, but as we anyway have an AWS machine for MSVC, it wasn't that challenging :wink:

File: .github/workflows/tests_linux.yml (hunk 2 above)
Comment (chaeyeunpark) [reply .github/workflows/tests_linux.yml:34]: Yes, Codecov raises this warning. I think this is the reason why CodeCov sometimes couldn't merge coverage data uploaded from the same commit has.

**Review by mlxd — APPROVED** [review]

**Pull request merged**